Support opening PDFs with Zotero to import

AbeJellinek opened the pull request:
Fixes #1169, needs zotero/zotero-standalone-build#90. Automatically attempts to recognize imported PDFs.

AbeJellinek committed:
Support opening PDFs with Zotero to import

dstillman commented on components/zotero-service.js:
```diff
@@ -612,42 +612,49 @@ ZoteroCommandLineHandler.prototype = {
+					else if (file.leafName.substr(-4).toLowerCase() === ".pdf") {
+						let item = await Zotero.Attachments.importFromFile({
+							file,
+							libraryID: Zotero.Libraries.userLibraryID
```
Should we have a dialog with a library drop-down, defaulting to the currently selected library? It's sort of weird if you're in another library and open a PDF via Zotero and then the PDF goes into your personal library.

dstillman replied:
Arguably we should use that for metadata files too. If you only have a personal library, it would skip the prompt.

AbeJellinek replied:
Added a unified dialog. It looks like this:

<img width="363" alt="image" src="https://user-images.githubusercontent.com/1770299/156023137-1a869cd6-09ea-45d4-96f7-af864a2eb070.png">

AbeJellinek committed:
Prompt for library when importing

This makes behavior consistent between PDF and translator imports. A new
dialog prompts for a target library (defaulting to the current one) and
whether to add to a new collection. The behavior is:

- If the user checks the "Add to new collection" box, the imported
  item(s) are added to a new collection within the target library.
- If the user doesn't check the box and the target library is the
  currently selected library, the item(s) are added to the currently
  selected collection.
- Otherwise, items are added to the root of the target library.

dstillman commented on chrome/content/zotero/selectImportTargetDialog.js:
```diff
@@ -0,0 +1,56 @@
+		= Zotero.getString('ingester.importFile.text', filename);
+
+	let menulist = document.getElementById('select-target-menu');
+	let libraries = Zotero.Libraries.getAll().filter(lib => lib.editable);
```
Also has to be `filesEditable` for PDFs

dstillman commented on chrome/locale/en-US/zotero/zotero.properties:
```diff
@@ -634,9 +634,9 @@ ingester.scrapeErrorDescription	= An error occurred while saving this item. Chec
-ingester.importFile.text	= Do you want to import the file "%S"?
-ingester.importFile.intoNewCollection = Import into new collection
+ingester.importFile.title				= Import File
+ingester.importFile.text				= Import “%S” to:
```
`to` → `into`

dstillman commented on components/zotero-service.js:
```diff
@@ -612,42 +612,82 @@ ZoteroCommandLineHandler.prototype = {
+					// Perform import in front window
+					var wm = Components.classes["@mozilla.org/appshell/window-mediator;1"]
+						.getService(Components.interfaces.nsIWindowMediator);
+					var browserWindow = wm.getMostRecentWindow("navigator:browser");
```
→ `Services.wm.getMostRecentWindow`

dstillman reviewed (changes requested):
I can't remember exactly how importing worked previously — if you didn't check `Import into new collection`, did it import into the currently selected collection or the root?

Even if the former, it's weirder with the addition of the library drop-down, because now you get different behavior based on whether you pick the current library (in which case it saves to the collection you happen to have selected behind the dialog) or another library (in which case it saves to the library root). There's no guarantee that people will have even been using Zotero recently, so I think it would be easy to accidentally import into a random collection just because you wanted something in the current library.

Maybe a better approach would be to use the full target drop-down, with the current collection selected by default. Bonus: if the current collection wasn't editable, you could then get immediate feedback on the fallback location (i.e., library root or personal library).

AbeJellinek commented:
The former, I think. I agree that it was/is weird. A target dropdown seems like a good approach. Should it have a “New Collection” option under each library that creates an automatically named new collection for the items, or should “New Collection…” open a dialog that prompts for a name (maybe with the automatic name as the default)?

dstillman commented:
> should “New Collection…” open a dialog that prompts for a name (maybe with the automatic name as the default)

I think this? The default name is pretty unlikely to be what you want, and it's just an extra Return if you don't care.

dstillman commented:
Oh, except this is for importing BibTeX/RIS/etc. too, in which case the auto-generated name might be fine. But still, it's just a Return. So I think show the dialog.

AbeJellinek committed:
Improve createMenuForTarget signature & behavior

- Made the clickable submenu behavior that itemPane.js implemented for
  feed items automatic.
- Event propagation is always stopped.
- Added a 'menusOnTopLevel' option in case the caller wants to ensure
  that every top-level node gets a menu, even if it has no children.
  This is useful for appending "New Collection..." menu items, for
  instance.

AbeJellinek committed:
Use target menu for imports, allow creating new collections

AbeJellinek committed:
Remove uses of new collection checkbox/pref

dstillman commented:
We should select the selected collection after import, and if it's a single item select the item too. Right now I believe it's just selecting the item in the current collection or library root if you select a collection other than the current one.

AbeJellinek committed:
Fix Add to Collection menu

AbeJellinek committed:
Convert _finishImport to ES6 async

Fixes error when onBeforeImport doesn't return a promise.

AbeJellinek committed:
Select target collection and single item after import

AbeJellinek committed:
Don't show target collection if recursiveCollections disabled

AbeJellinek commented:
OK, done.
